libcasr/src/cluster.rs

```diff
@@ -0,0 +1,340 @@
+//! Provides API's for cluster manipulating.
+use crate::error::*;
+use crate::stacktrace::*;
+
+use core::f64::MAX;
+use std::collections::HashMap;
+use std::path::PathBuf;
+
+/// Represents the information about CASR report
+pub type ReportInfo = (PathBuf, (Stacktrace, String));
+
+/// Relation between a CASR report and a cluster
+pub enum Relation {
+    /// The CASR report is a duplicate of one from cluster
+    Dup,
+    /// The CASR report is "inside" the cluster with some proximity measure
+    Inner(f64),
+    /// The CASR report is "outside" the cluster
+    Outer,
+}
+
+/// Structure provides an abstraction for cluster with CASR reports
+pub struct Cluster {
+    /// Cluster number
+    pub number: usize,
+    /// Cluster report paths
+    paths: Vec<PathBuf>,
+    /// Cluster report stacktraces
+    stacktraces: Vec<Stacktrace>,
+    /// Cluster diameter
+    diam: Option<f64>,
+    /// Cluster report crashlines
+    crashlines: HashMap<String, usize>,
+}
+
+impl Cluster {
+    /// Create new `Cluster`
+    pub fn new(
+        number: usize,
+        paths: Vec<PathBuf>,
+        stacktraces: Vec<Stacktrace>,
+        crashlines: Vec<String>,
+    ) -> Self {
+        let mut unique_crashlines: HashMap<String, usize> = HashMap::new();
+        for (i, crashline) in crashlines.into_iter().enumerate() {
+            unique_crashlines.insert(crashline, i);
+        }
+        Cluster {
+            number,
+            paths,
+            stacktraces,
+            diam: None,
+            crashlines: unique_crashlines,
+        }
+    }
+    /// Get CASR report paths
+    pub fn paths(&self) -> &Vec<PathBuf> {
+        &self.paths
+    }
+    /// Get CASR report stactraces
+    pub fn stacktraces(&self) -> &Vec<Stacktrace> {
+        &self.stacktraces
+    }
```

Comment:
Can we just make the fields public and remove these getters?

Reply:
No, we can't, cuz it logically connected with `crashlines` field